parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java

```diff
@@ -0,0 +1,531 @@
+      } catch (IllegalAccessException e) {
+        throw new DirectCodecPool.ParquetCompressionCodecException(e);
+      } catch (InvocationTargetException e) {
+        throw new DirectCodecPool.ParquetCompressionCodecException(e);
+      }
+      output.position(uncompressedSize);
+    }
+
+    @Override
+    protected void release() {
+      compressedBuffer = DirectCodecFactory.this.release(compressedBuffer);
+      uncompressedBuffer = DirectCodecFactory.this.release(uncompressedBuffer);
+      DirectCodecPool.INSTANCE.returnDirectDecompressor(decompressor);
+      extraDecompressor.release();
+    }
+
+  }
+
+  public class NoopDecompressor extends BytesDecompressor {
+
+    @Override
+    public void decompress(ByteBuffer input, int compressedSize, ByteBuffer output, int uncompressedSize)
+        throws IOException {
+      Preconditions.checkArgument(compressedSize == uncompressedSize,
+          "Non-compressed data did not have matching compressed and uncompressed sizes.");
+      output.clear();
+      output.put((ByteBuffer) input.duplicate().position(0).limit(compressedSize));
+    }
+
+    @Override
+    public BytesInput decompress(BytesInput bytes, int uncompressedSize) throws IOException {
+      return bytes;
+    }
+
+    @Override
+    protected void release() {}
+
+  }
+
+  public class SnappyDecompressor extends BytesDecompressor {
+
+    private HeapBytesDecompressor extraDecompressor;
+    public SnappyDecompressor() {
+      this.extraDecompressor = new HeapBytesDecompressor(CompressionCodecName.SNAPPY);
+    }
+
+    @Override
+    public BytesInput decompress(BytesInput bytes, int uncompressedSize) throws IOException {
+      return extraDecompressor.decompress(bytes, uncompressedSize);
+    }
+
+    @Override
+    public void decompress(ByteBuffer src, int compressedSize, ByteBuffer dst, int uncompressedSize) throws IOException {
+      dst.clear();
+      int size = Snappy.uncompress(src, dst);
+      dst.limit(size);
+    }
+
+    @Override
+    protected void release() {}
+  }
+
+  public class SnappyCompressor extends BytesCompressor {
+
+    // TODO - this outgoing buffer might be better off not being shared, this seems to
+    // only work because of an extra copy currently happening where this interface is
+    // be consumed
+    private ByteBuffer incoming;
+    private ByteBuffer outgoing;
+
+    /**
+     * Compress a given buffer of bytes
+     * @param bytes
+     * @return
+     * @throws IOException
+     */
+    @Override
+    public BytesInput compress(BytesInput bytes) throws IOException {
+      int maxOutputSize = Snappy.maxCompressedLength((int) bytes.size());
+      ByteBuffer bufferIn = bytes.toByteBuffer();
+      outgoing = ensure(outgoing, maxOutputSize);
+      final int size;
+      if (bufferIn.isDirect()) {
+        size = Snappy.compress(bufferIn, outgoing);
+      } else {
+        // Snappy library requires buffers be direct
```

Comment:
I think it'd be better to put a restriction on this class to only allow for direct buffer allocators to be passed in the constructor. Do you think it would make sense to add a method isDirect to the BufferAllocator interface or just move the DirectBufferAllocator to be an interface with docs to say that the buffers allocated from it must be direct? Drill has an alternative implementation of this interface to work it into our account of direct memory. We could just make this drill class a subclass of DirectBufferAlocator (so I could put the same docs in the DirectBufferaAllocator class), but then we wouldn't be using any of the content of the class for Drill, we would just be using like an interface anyway.